doc/man1/openssl-pkeyutl.pod.in

```diff
@@ -133,22 +142,37 @@ The input is a certificate containing a public key.
 =item B<-rev>
 
 Reverse the order of the input buffer. This is useful for some libraries
-(such as CryptoAPI) which represent the buffer in little endian format.
+(such as CryptoAPI) which represent the buffer in little-endian format.
 This cannot be used in conjunction with B<-rawin>.
 
 =item B<-sign>
 
-Sign the input data (which must be a hash) and output the signed result. This
-requires a private key.
+Sign the input data and output the signed result. This requires a private key.
+Using a message digest operation along with this is recommended -
+see the B<-rawin> and B<-digest> options for details.
+Otherwise, the input data given with the B<-in> option is assumed to already
+be a digest, but this may then require an additional B<-pkeyopt> C<digest:>I<md>
+in some cases (e.g., RSA with the default PKCS#1 padding mode).
+Even for other algorithms like ECDSA, where the additional B<-pkeyopt> option
+does not affect signature output, it is recommended, as it enables
+checking that the input length is consistent with the intended digest.
 
 =item B<-verify>
 
-Verify the input data (which must be a hash) against the signature file and
-indicate if the verification succeeded or failed.
+Verify the input data against the signature given with the B<-sigfile> option
+and indicate if the verification succeeded or failed.
+The input data given with the B<-in> option must be a hash value
+unless the B<-rawin> option is specified or implied.
+With raw data, when a digest algorithm is applicable, though it may be inferred
+from the signature or take a default value, it should also be specified.
 
 =item B<-verifyrecover>
 
-Verify the input data (which must be a hash) and output the recovered data.
+Verify the input data and output the recovered data (signature payload). For
+example, in the case of RSA PKCS#1 the recovered data is the B<EMSA-PKCS-v1_5>
+DER encoding of the digest algorithm OID and value as specified in
+L<RFC8017 Section 9.2|https://datatracker.ietf.org/doc/html/rfc8017#section-9.2>.
+The input data given with the B<-in> option should be a hash value.
```

Comment:
```suggestion
The input data given with the B<-in> option must be a hash value.
```

Reply:
Done.

Although for the `-verifyrecover` option, as well as for `-sign` and `-verify`,
there could still be low-level use cases where data is intended to get signed directly (without being hashed).

Reply:
If that that's possible in principle, last I checked `-verifyrecover` refuses to accept raw input.

Reply:
When the input is not large, it appears to work - just tried with examples from the doc:
```
echo "1234" > file
ln -s test/certs/root-key.pem key.pem
openssl pkeyutl -sign -in file -inkey key.pem -out sig
openssl pkeyutl -verifyrecover -in sig -inkey key.pem
# yields: 1234
```

Reply:
> When the input is not large, it appears to work - just tried with examples from the doc:
> 
> ```
> echo "1234" > file
> ln -s test/certs/root-key.pem key.pem
> openssl pkeyutl -sign -in file -inkey key.pem -out sig
> openssl pkeyutl -verifyrecover -in sig -inkey key.pem
> # yields: 1234
> ```

This "works" because "1234" is assumed to be the digest. You did not use when signing, nor with "verifyrecover", so with RSA signatures of short "undigested" input, the decrypted signature payload is "1234", but this is not exactly a useful facility.